### PR TITLE
Add Kendall Rank Correlation Coefficient (Kendall's tau)

### LINKED ISSUE #1206: Add Kendall's Rank Correlation
## 🚀 Feature

A `Metric` that computes Kendall's Rank Correlation (https://en.wikipedia.org/wiki/Kendall_rank_correlation_coefficient).

### Motivation

This metric would be useful for many regressive tasks.

### Pitch

Add a new `Metric` that computes Kendall's Rank Correlation - specifically its `tau` values - possibly using SciPy's implementation as a reference (https://web.archive.org/web/20181008171919/https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kendalltau.html).

### Alternatives

One currently can compute this quantity using SciPy. However, it would be desirable to compute it directly using PyTorch.

### Additional context

Whoever implements this new `Metric` deserves at least a (virtual) high-five! :hand: 

Note that the `tau` values associated with this metric can be between -1 and 1, with 1 indicating a perfect relationship between a pair of input variables. However, according to this source (https://www.statisticshowto.com/kendalls-tau/), any negative signs on `tau` can safely be discarded.

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?

Fixes #1206 

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/torchmetrics/functional/regression/kendall.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    preds_ties_p1: Optional[Tensor],
+    preds_ties_p2: Optional[Tensor],
+    target_ties: Optional[Tensor],
+    target_ties_p1: Optional[Tensor],
+    target_ties_p2: Optional[Tensor],
+    variant: _MetricVariant,
+    alternative: Optional[_TestAlternative],
+) -> Tensor:
+    """Calculate p-value for Kendall's tau from metric metadata."""
+    normal_dist = torch.distributions.normal.Normal(torch.tensor([0.0]), torch.tensor([1.0]))
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/regression/kendall.py:210]: should this be moved to the same device as `con_min_dis_pairs`?

File: src/torchmetrics/functional/regression/kendall.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    for i in range(x.shape[0]):
+        y[i] = y[i][perm[i]]
+    return x.T, y.T
+
+
+def _count_concordant_pairs(preds: Tensor, target: Tensor) -> Tensor:
+    """Count a total number of concordant pairs in given sequences."""
+
+    def _concordant_element_sum(x: Tensor, y: Tensor, i: int) -> Tensor:
+        return torch.logical_and(x[i] < x[(i + 1) :], y[i] < y[(i + 1) :]).sum(0).unsqueeze(0)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/regression/kendall.py:85]: nit: could we not have nested functions?

File: src/torchmetrics/functional/regression/kendall.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+class _MetricVariant(EnumStr):
+    """Enumerate for metric variants."""
+
+    A = "a"
+    B = "b"
+    C = "c"
+
+    @classmethod
+    def from_str(cls, value: str) -> Optional["EnumStr"]:
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/regression/kendall.py:34]: could this `from_str` be implemented in the base class?
it seems to be in every subclass with only slight variations

File: src/torchmetrics/functional/regression/kendall.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+        tau_b = \frac{C - D}{\sqrt{(C + D + T_{preds}) * (C + D + T_{target})}}
+
+        tau_c = 2 * \frac{C - D}{n ** 2 * \frac{m - 1}{m}}
+
+    where :math:`C` is represents concordant pairs, :math:`D` stands for discordant pairs and :math:`T` represents
+    a total number of ties. Definition according to `The Treatment of Ties in Ranking Problems`_.
+
+    Args:
+        preds: Ordered sequence of data
+        target: Ordered sequence of data
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/regression/kendall.py:351]: should they really be ordered sequences of data?
the data in the examples does not seem to be

File: src/torchmetrics/functional/regression/kendall.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        >>> preds = torch.tensor([2.5, 0.0, 2, 8])
+        >>> kendall_rank_corrcoef(preds, target)
+        tensor(0.3333)
+
+    Example (multi output regression):
+        >>> from torchmetrics.functional.regression import kendall_rank_corrcoef
+        >>> target = torch.tensor([[3, -0.5], [2, 1]])
+        >>> preds = torch.tensor([[2.5, 0.0], [2, 8]])
+        >>> kendall_rank_corrcoef(preds, target)
+        tensor([1., 1.])
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/regression/kendall.py:379]: could we include an third example with `t_test=True` to show that the metric can output an second value

File: src/torchmetrics/functional/regression/kendall.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    if preds.ndim > 2 or target.ndim > 2:
+        raise ValueError(
+            f"Expected both predictions and target to be either 1- or 2-dimensional tensors,"
+            f" but got {target.ndim} and {preds.ndim}."
+        )
+    if (num_outputs == 1 and preds.ndim != 1) or (num_outputs > 1 and num_outputs != preds.shape[1]):
+        raise ValueError(
+            f"Expected argument `num_outputs` to match the second dimension of input, but got {num_outputs}"
+            f" and {preds.shape[1]}."
+        )
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/regression/kendall.py:262]: maybe at this point it is worth refactoring this into an utility function when, since it is repeated between all three correlation metrics

File: src/torchmetrics/regression/kendall.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    Example (multi output regression):
+        >>> import torch
+        >>> from torchmetrics.regression import KendallRankCorrCoef
+        >>> target = torch.tensor([[3, -0.5], [2, 1]])
+        >>> preds = torch.tensor([[2.5, 0.0], [2, 8]])
+        >>> kendall = KendallRankCorrCoef(num_outputs=2)
+        >>> kendall(preds, target)
+        tensor([1., 1.])
+    """
```
Comment (SkafteNicki) [inline src/torchmetrics/regression/kendall.py:77]: again an extra example would be nice showcasing `t_test` argument

**Review by SkafteNicki — APPROVED** [review]

File: src/torchmetrics/functional/regression/kendall.py (hunk 3 above)
Comment (stancld) [reply src/torchmetrics/functional/regression/kendall.py:35]: I will check this what I can do it here

File: tests/unittests/regression/test_kendall.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+import torch
+from scipy.stats import kendalltau
+
+from torchmetrics.functional.regression.kendall import kendall_rank_corrcoef
+from torchmetrics.regression.kendall import KendallRankCorrCoef
+from torchmetrics.utilities.imports import _compare_version
+from unittests.helpers import seed_all
+from unittests.helpers.testers import BATCH_SIZE, EXTRA_DIM, NUM_BATCHES, MetricTester
+
+_SCIPY_GREATER_EQUAL_1_8 = _compare_version("scipy", operator.ge, "1.8.0")
```
Comment (SkafteNicki) [inline tests/unittests/regression/test_kendall.py:28]: Is there a reason we are not fixing scipy to be >=1.8.0?
Should not be a problem when only used for testing.

File: src/torchmetrics/regression/kendall.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+    where :math:`C` represents concordant pairs, :math:`D` stands for discordant pairs, :math:`n` is a total number
+    of observations and :math:`m` is a ``min`` of unique values in ``preds`` and ``target`` sequence.
+
+    Definitions according to Definition according to `The Treatment of Ties in Ranking Problems`_.
+
+    Forward accepts
+
+    - ``preds``: Sequence of data
+    - ``target``: Sequence of data
```
Comment (SkafteNicki) [inline src/torchmetrics/regression/kendall.py:55]:
```suggestion
    - ``preds`` (float tensor): Sequence of data of either shape ``(N,)`` or ``(N,d)``
    - ``target`` (float tensor): Sequence of data of either shape ``(N,)`` or ``(N,d)``
```

File: src/torchmetrics/regression/kendall.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+        self.add_state("preds", [], dist_reduce_fx="cat")
+        self.add_state("target", [], dist_reduce_fx="cat")
+
+    def update(self, preds: Tensor, target: Tensor) -> None:
+        """Update variables required to compute Kendall rank correlation coefficient.
+
+        Args:
+            preds: Sequence of data
+            target: Sequence of data
```
Comment (SkafteNicki) [inline src/torchmetrics/regression/kendall.py:140]:
```suggestion
            preds: Sequence of data of either shape ``(N,)`` or ``(N,d)``
            target: Sequence of data of either shape ``(N,)`` or ``(N,d)``
```

File: src/torchmetrics/functional/regression/kendall.py [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+        tau_c = 2 * \frac{C - D}{n^2 * \frac{m - 1}{m}}
+
+    where :math:`C` represents concordant pairs, :math:`D` stands for discordant pairs, :math:`n` is a total number
+    of observations and :math:`m` is a ``min`` of unique values in ``preds`` and ``target`` sequence.
+
+    Definitions according to Definition according to `The Treatment of Ties in Ranking Problems`_.
+
+    Args:
+        preds: Sequence of data
+        target: Sequence of data
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/regression/kendall.py:370]:
```suggestion
        preds: Sequence of data of either shape ``(N,)`` or ``(N,d)``
        target: Sequence of data of either shape ``(N,)`` or ``(N,d)``
```

**Comment by SkafteNicki** [comment]
@Borda can you help land this. Code works as expected but currently one general mypy check is failing with:
```
error: --install-types failed (no mypy cache directory)
```

**Comment by Borda** [comment]
> @Borda can you help land this. Code works as expected but currently one general mypy check is failing with:
> 
> ```
> error: --install-types failed (no mypy cache directory)
> ```

hmmm never seen this before :/ let me check on MOnday :rabbit:

**Comment by stancld** [comment]
@SkafteNicki Do we wanna keep the last to commits adding new modules to `mypy` ignore list?

**Comment by SkafteNicki** [comment]
> @SkafteNicki Do we wanna keep the last to commits adding new modules to `mypy` ignore list?

Nope, have removed them again :)

**Comment by SkafteNicki** [comment]
> > @Borda can you help land this. Code works as expected but currently one general mypy check is failing with:
> > ```
> > error: --install-types failed (no mypy cache directory)
> > ```
> 
> hmmm never seen this before :/ let me check on MOnday 🐰

The weird error was fixed in [fefe505](https://github.com/Lightning-AI/metrics/pull/1271/commits/fefe50529c6e75169f3b69b84b2c7900b1495f70).
@stancld could you take care of the remaining mypy typing issues, mostly have to do with the use `EnumStr`

**Comment by stancld** [comment]
@Borda Should be ready for a final review :]

**Review by Borda — APPROVED** [review]
